library/alloc/src/collections/btree/node.rs

```diff
@@ -528,28 +508,28 @@ impl<'a, K: 'a, V: 'a> NodeRef<marker::Mut<'a>, K, V, marker::Internal> {
     }
 }
 
-impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Immut<'a>, K, V, Type> {
+impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Mut<'a>, K, V, Type> {
     /// Exposes the entire key storage area in the node,
     /// regardless of the node's current length,
     /// having exclusive access to the entire node.
-    unsafe fn key_area(self) -> &'a [MaybeUninit<K>] {
-        Self::as_leaf(&self).keys.as_slice()
+    unsafe fn key_area(mut self) -> &'a [MaybeUninit<K>] {
+        Self::as_leaf_mut(&mut self).keys.as_slice()
     }
 
     /// Exposes the entire value storage area in the node,
     /// regardless of the node's current length,
     /// having exclusive access to the entire node.
-    unsafe fn val_area(self) -> &'a [MaybeUninit<V>] {
-        Self::as_leaf(&self).vals.as_slice()
+    unsafe fn val_area(mut self) -> &'a [MaybeUninit<V>] {
+        Self::as_leaf_mut(&mut self).vals.as_slice()
     }
 }
 
-impl<'a, K: 'a, V: 'a> NodeRef<marker::Immut<'a>, K, V, marker::Internal> {
+impl<'a, K: 'a, V: 'a> NodeRef<marker::Mut<'a>, K, V, marker::Internal> {
     /// Exposes the entire storage area for edge contents in the node,
     /// regardless of the node's current length,
     /// having exclusive access to the entire node.
-    unsafe fn edge_area(self) -> &'a [MaybeUninit<BoxedNode<K, V>>] {
-        Self::as_internal(&self).edges.as_slice()
+    unsafe fn edge_area(mut self) -> &'a [MaybeUninit<BoxedNode<K, V>>] {
+        Self::as_internal_mut(&mut self).edges.as_slice()
```

Comment:
It seems quite odd to require unique access to get a read-only slice of the edges -- can you say more about the use case here?

Reply:
The change is not to require unique access (to the tree), we admit that the caller has it and needs to have it. We don't get you a slice of edges, but a slice of containers of edges. If you only have shared access to the tree, the only thing you could ever do with those containers is to read the initialized ones, and we already have better methods for that.

The actual use case is to unify the various methods that access `MaybeUninit`s in another PR.